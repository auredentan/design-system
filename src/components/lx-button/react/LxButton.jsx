import React from 'react';
import classNames from 'classnames';

import { handleBasicClasses } from '../../../core/utils';
import '../style/lx-button.scss';

const LxButtonRoot = ({ href, children, ...props }) => {
    return href ? (
        <a href={href} {...props}>
            {children}
        </a>
    ) : (
        <button {...props}>{children}</button>
    );
};

const LxButton = ({
    children,
    className,
    color = 'primary',
    variant = 'primary',
    size = 'm',
    theme = 'light',
    ...props
}) => (
    <LxButtonRoot
        className={classNames(className, handleBasicClasses({ color, size, theme, variant, prefix: 'lx-button' }))}
        {...props}
    >
        {children}
    </LxButtonRoot>
);

export { LxButton };
