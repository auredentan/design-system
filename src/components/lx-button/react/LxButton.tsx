import React from 'react';
import classNames from 'classnames';

import '../style/lx-button.scss';

import { handleBasicClasses } from '../../../core/utils';
import { Color, Theme, Size, Variant } from 'components';

interface LxButtonRootProps {
    /** Basic react `className` prop. */
    className?: string;
    /** The `href` to reach if there is one. */
    href?: string;
    /** The `target` html property. */
    target?: string;
}

/**
 * Conditionnaly adds `a` or `button` html tag whether there is an `href` attribute or not.
 */
const LxButtonRoot: React.FC<LxButtonRootProps> = ({ href, children, target, className, ...props }) => {
    return href ? (
        <a href={href} target={target} className={className} {...props}>
            {children}
        </a>
    ) : (
        <button className={className} {...props}>
            {children}
        </button>
    );
};

export interface LxButtonProps extends LxButtonRootProps {
    /** The button color which must be defined by `lx-button--${color}` css class. */
    color?: Color;
    /** The button size which must be defined by `lx-button--${size}` css class. */
    size?: Size;
    /** The button theme which must be defined by `lx-button--${theme}` css class. */
    theme?: Theme;
    /** The button variant which must be defined by `lx-type--${variant}` css class. */
    variant?: Variant;
}

/**
 * Displays a button. If `href` property is set, it will display an `a` tag, if not
 * it will use `button` tag instead.
 */
export const LxButton: React.FC<LxButtonProps> = ({
    children,
    className,
    color = 'primary',
    variant = 'primary',
    size = 'm',
    theme = 'light',
    ...props
}) => (
    <LxButtonRoot
        className={classNames(className, handleBasicClasses({ color, size, theme, variant, prefix: 'lx-button' }))}
        {...props}
    >
        {children}
    </LxButtonRoot>
);
