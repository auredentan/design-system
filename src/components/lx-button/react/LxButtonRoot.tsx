import React, { Children } from 'react';

import isEmpty from 'lodash/isEmpty';

/////////////////////////////

/**
 * Defines the props of the <LxButtonRoot> component.
 */
interface ILxButtonRootProps {
    /**
     * The `href` to reach if there is one.
     */
    href?: string;

    /**
     * The `target` to open the `href` into.
     */
    target?: string;

    /**
     * Any other supported prop for a HTML element.
     * E.g. classNames, onClick, disabled, ...
     */
    [propName: string]: any;
}
type LxButtonRootProps = ILxButtonRootProps;

/////////////////////////////

/**
 * The root of the <LxButton> component.
 * Conditionnaly adds a `<a>` or a `<button>` HTML tag whether there is an `href` attribute or not.
 *
 * @return {JSX.Element} The <LxButton> root component.
 */
const LxButtonRoot: React.FC<LxButtonRootProps> = ({
    children,
    href,
    target,
    ...props
}: LxButtonRootProps): JSX.Element => {
    const childrenCount: number = Children.count(children);
    if (childrenCount === 0) {
        throw new Error('<LxButtonRoot> must have at least 1 child (got 0)!');
    }

    if (isEmpty(href)) {
        return <button {...props}>{children}</button>;
    }

    return (
        <a href={href} target={target} {...props}>
            {children}
        </a>
    );
};
LxButtonRoot.displayName = 'LxButtonRoot';

/////////////////////////////

export { LxButtonRoot, LxButtonRootProps };
