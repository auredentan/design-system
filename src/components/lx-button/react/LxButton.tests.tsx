import { LxButtonProps } from './LxButton';

/////////////////////////////

import React, { Fragment } from 'react';

import { shallow, ShallowWrapper } from 'enzyme';

import { LxButton, LxIcon } from 'LumX';
import { ICommonSetup } from 'LumX/core/testing/utils.tests';
import { mdiPlus } from 'LumX/icons';

/////////////////////////////

interface ISetup extends ICommonSetup {
    props: LxButtonProps;
}

/////////////////////////////

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param  {LxButtonProps} props The props to use to override the default props of the component.
 * @return {ISetup}        An object with the props, the component wrapper and some shortcut to some element inside of
 *                         the component.
 */
const setup = ({ ...propsOverrides }: LxButtonProps): ISetup => {
    const props: LxButtonProps = {
        ...propsOverrides,
    };

    const wrapper: ShallowWrapper = shallow(<LxButton {...props} />);

    return {
        props,
        wrapper,
    };
};

describe(`<${LxButton.displayName}>`, () => {
    // 1. Test render via snapshot (default state of component).
    describe('Snapshots', () => {
        it('should render correctly a text label', () => {
            const children: React.ReactNode = 'Label';

            const { wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render correctly a <span> label', () => {
            const children: React.ReactNode = <span>label</span>;

            const { wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render correctly an icon and a text label', () => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    label
                </Fragment>
            );

            const { wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render correctly an icon and a <span> label', () => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    <span>label</span>
                </Fragment>
            );

            const { wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render correctly a text label and an icon', () => {
            const children: React.ReactNode = (
                <Fragment>
                    label
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            const { wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render correctly a <span> label and an icon', () => {
            const children: React.ReactNode = (
                <Fragment>
                    <span>label</span>
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            const { wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render correctly two icons and a text label', () => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    label
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            const { wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render correctly two icons and a <span> label', () => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    <span>label</span>
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            const { wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();
        });

        it("should render correctly an icon button with the 'button' `variant`", () => {
            const children: React.ReactNode = <LxIcon icon={mdiPlus} />;

            const { wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();
        });

        it("should render correctly an icon button with the 'icon' `variant`", () => {
            const children: React.ReactNode = <LxIcon icon={mdiPlus} />;

            const { wrapper }: ISetup = setup({ children, variant: 'icon' });
            expect(wrapper).toMatchSnapshot();
        });
    });

    /////////////////////////////

    // 2. Test defaultProps value and important props custom values.

    /////////////////////////////

    // 3. Test events.

    /////////////////////////////

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Condition', () => {
        beforeEach(() => {
            try {
                // If `console.warn` has been mocked at least one, this exists. So disable TS here.
                // @ts-ignore
                global.console.warn.mockClear();
            } catch (exception) {
                // Nothing to do here.
            }
        });

        it(`should fail when no child is given`, () => {
            expect(() => {
                // We know that children is mandatory, but we want to test the error validation. So disable TS here.
                // @ts-ignore
                setup({});
            }).toThrowErrorMatchingSnapshot();
        });

        it("should fail when more than 3 children are given in the 'button' `variant`", () => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    <span>label</span>
                    <span>label 2</span>
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            expect(() => {
                // We know that children is mandatory, but we want to test the error validation. So disable TS here.
                // @ts-ignore
                setup({ children });
            }).toThrowErrorMatchingSnapshot();
        });

        it(`should fail when anything else than a text, <span> or <${
            LxIcon.displayName
        }> is given as child in the 'button' \`variant\``, () => {
            const children: React.ReactNode = <div>toto</div>;

            expect(() => {
                setup({ children });
            }).toThrowErrorMatchingSnapshot();
        });

        it("should fail when more than 1 text or <span> child is given in the 'button' `variant`", () => {
            let children: React.ReactNode = (
                <Fragment>
                    <span>label</span>
                    label 2
                </Fragment>
            );

            expect(() => {
                // We know that children is mandatory, but we want to test the error validation. So disable TS here.
                // @ts-ignore
                setup({ children });
            }).toThrowErrorMatchingSnapshot();

            children = (
                <Fragment>
                    <span>label</span>
                    <span>label 2</span>
                </Fragment>
            );

            expect(() => {
                // We know that children is mandatory, but we want to test the error validation. So disable TS here.
                // @ts-ignore
                setup({ children });
            }).toThrowErrorMatchingSnapshot();
        });

        it(`should fail when there are 2 or more <${LxIcon.displayName}> in a row in the 'button' \`variant\``, () => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            expect(() => {
                // We know that children is mandatory, but we want to test the error validation. So disable TS here.
                // @ts-ignore
                setup({ children });
            }).toThrowErrorMatchingSnapshot();
        });

        it("should fail when more than 1 child is given in the 'icon' `variant`", () => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            expect(() => {
                // We know that children is mandatory, but we want to test the error validation. So disable TS here.
                // @ts-ignore
                setup({ children, variant: 'icon' });
            }).toThrowErrorMatchingSnapshot();
        });

        it(`should fail when anything else than a <${
            LxIcon.displayName
        }> is given as child in the 'icon' \`variant\``, () => {
            const children: React.ReactNode = 'Toto';

            expect(() => {
                setup({ children, variant: 'icon' });
            }).toThrowErrorMatchingSnapshot();
        });

        it("should warn the user when rendering an icon button with the 'button' `variant`", () => {
            global.console.warn = jest.fn();

            const children: React.ReactNode = <LxIcon icon={mdiPlus} />;

            setup({ children });
            expect(global.console.warn).toHaveBeenCalled();
        });

        it("should not warn the user when rendering an icon button with the 'icon' `variant`", () => {
            global.console.warn = jest.fn();

            const children: React.ReactNode = <LxIcon icon={mdiPlus} />;

            setup({ children, variant: 'icon' });
            expect(global.console.warn).not.toHaveBeenCalled();
        });
    });

    /////////////////////////////

    // 5. Test state.
});
