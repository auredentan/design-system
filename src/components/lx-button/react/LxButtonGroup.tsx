import React, { Children } from 'react';

import { isElementOfType } from 'LumX/core/react/utils';

import { CLASSNAME as LXBUTTON_CLASSNAME, LxButton } from './LxButton';

/////////////////////////////
/**
 * Defines the props of the component
 */
interface IProps {
    /**
     * Basic react `children` prop.
     */
    children: React.ReactNode;
}

/////////////////////////////
//                         //
//    Private functions    //
//                         //
/////////////////////////////

/**
 * Validate the <LxButtonGroup> component props and children.
 * Also, sanitize, cleanup and format the children and return the processed ones.
 *
 * @param  {IProps}  props The children and props of the <LxButtonGroup> component.
 * @return {React.ReactNode} The processed children of the component.
 */
function _validate({ children }: IProps): React.ReactNode {
    Children.forEach(
        children,
        (child: any): void => {
            if (!isElementOfType(child, LxButton)) {
                throw new Error(`You can only have <LxButton>s child in a <LxButtonGroup> (got ${child.type})!`);
            }
        },
    );

    return children;
}

/////////////////////////////

/**
 * Displays a group of <LxButton>s.
 *
 * @see {@link LxButton} for more information on <LxButton>.
 *
 * @return {JSX.Element} The <LxButtonGroup> component.
 */
const LxButtonGroup: React.FC<IProps> = ({ children }: IProps): JSX.Element => {
    _validate({ children });

    return <div className={`${LXBUTTON_CLASSNAME}-group`}>{children}</div>;
};
LxButtonGroup.displayName = 'LxButtonGroup';

/////////////////////////////

export { LxButtonGroup };
