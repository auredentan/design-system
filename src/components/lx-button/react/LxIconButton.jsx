import React, { Children, cloneElement } from 'react';
import classNames from 'classnames';

import { LxButton } from './LxButton';
import '../style/lx-button.scss';

export const LxIconButton = ({ children, ...props }) => (
    <LxButton className="lx-button--shape-circled" {...props}>
        {Children.map(children, (child) =>
            cloneElement({
                ...child,
                props: {
                    ...child.props,
                    className: classNames(child.props.className, 'lx-button__icon'),
                },
            }),
        )}
    </LxButton>
);
