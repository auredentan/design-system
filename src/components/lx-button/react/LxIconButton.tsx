import React, { Children, cloneElement } from 'react';

import classNames from 'classnames';

import '../style/lx-button.scss';

import { ILxButtonProps, LxButton } from './LxButton';

export const LxIconButton: React.FC<ILxButtonProps> = ({ children, ...props }) => (
    <LxButton className="lx-button--shape-circled" {...props}>
        {Children.map(children, (child: any) =>
            cloneElement(child, {
                className: classNames(child.props.className, 'lx-button__icon'),
            }),
        )}
    </LxButton>
);
