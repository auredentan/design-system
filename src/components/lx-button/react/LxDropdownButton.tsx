import React, { Fragment, useState } from 'react';

import classNames from 'classnames';

import isEmpty from 'lodash/isEmpty';
import isFunction from 'lodash/isFunction';

import { LxDropdown, LxIcon, LxIconButton } from 'LumX';
import { IGenericProps, validateComponent, ValidateParameters } from 'LumX/core/react/utils';
import { mdiMenuDown } from 'LumX/icons';

import {
    CLASSNAME as LXBUTTON_CLASSNAME,
    Color,
    Colors,
    Emphasis,
    Emphasises,
    LxButton,
    LxButtonProps,
    Size,
    Sizes,
    Theme,
    Themes,
} from './LxButton';
import { LxButtonGroup, LxButtonGroupProps } from './LxButtonGroup';

/////////////////////////////

enum Variants {
    button = 'button',
}
type Variant = Variants;

/////////////////////////////

/**
 * Defines the props of the component.
 */
interface IProps extends IGenericProps {
    /**
     * Contains the dropdown element to display when the icon is clicked.
     */
    dropdown?: JSX.Element;

    /**
     * Indicates if the label <LxButton> and the icon <LxDropdownButton> buttons are separated in the <LxButtonGroup>
     */
    isSplitted?: boolean;

    /**
     * The <LxDropdownButton> should never have the `variant` prop as this prop is forced to 'button' in the <LxButton>.
     */
    variant?: never;
}
type LxDropdownButtonProps = IProps & LxButtonProps & LxButtonGroupProps;

/////////////////////////////

/**
 * Define the types of the default props.
 */
interface ILxDropdownButtonDefaultPropsType extends Partial<LxDropdownButtonProps> {}

/////////////////////////////
//                         //
//    Public attributes    //
//                         //
/////////////////////////////

/**
 * The default class name and classes prefix for this component.
 *
 * @type {string}
 * @constant
 * @readonly
 */
const CLASSNAME: string = `${LXBUTTON_CLASSNAME}__dropdown`;

/**
 * The display name of the component.
 *
 * @type {string}
 * @constant
 * @readonly
 */
const COMPONENT_NAME: string = 'LxDropdownButton';

/**
 * The default value of props.
 *
 * @type {ILxDropdownButtonDefaultPropsType}
 * @constant
 * @readonly
 */
const DEFAULT_PROPS: ILxDropdownButtonDefaultPropsType = {
    isSplitted: false,
};

/////////////////////////////
//                         //
//    Private functions    //
//                         //
/////////////////////////////

/**
 * Globally validate the <LxDropdownButton> component before validating the children.
 *
 * @param {ValidateParameters} props The properties of the component.
 */
function _preValidate({ props }: ValidateParameters): void {
    if (isEmpty(props.dropdown)) {
        console.warn(
            `You haven't specified any dropdown for you <${COMPONENT_NAME}>. The dropdown toggle will not work until you provide one.`,
        );
    }

    if (isEmpty(props.variant)) {
        return;
    }

    console.warn(
        `You shouldn't pass the \`variant\` prop in a <${COMPONENT_NAME}> as it's forced to 'button' (got '${
            props.variant
        }')!`,
    );
}

/**
 * Validate the <LxDropdownButton> component props and children.
 * Also, sanitize, cleanup and format the children and return the processed ones.
 *
 * @param  {LxDropdownButtonProps} props The children and props of the <LxButton> component.
 * @return {React.ReactNode}       The processed children of the component.
 */
function _validate(props: LxDropdownButtonProps): React.ReactNode {
    return validateComponent(COMPONENT_NAME, {
        maxChildren: 2,
        preValidate: _preValidate,
        props,
    });
}

/////////////////////////////

/**
 * Displays an icon button.
 * It's like a <LxButton> but displays an icon instead of a label in the body of the button.
 *
 * @see {@link LxButton} for more information on <LxButton>.
 *
 * @return {JSX.Element} The <LxDropdownButton> component.
 */
const LxDropdownButton: React.FC<LxDropdownButtonProps> = ({
    children,
    className = '',
    dropdown,
    isSplitted = DEFAULT_PROPS.isSplitted,
    ...props
}: LxDropdownButtonProps): JSX.Element => {
    const [isDropdownOpened, setIsDropdownOpened]: [boolean, (isDropdownOpened: boolean) => void] = useState(false);

    const newChildren = _validate({ children, dropdown, isSplitted, ...props });

    const openDropdown: (evt: React.MouseEvent<HTMLElement>) => boolean | void = (
        evt: React.MouseEvent<HTMLElement>,
    ): boolean | void => {
        setIsDropdownOpened(!isDropdownOpened);

        if (!isSplitted && isFunction(props.onClick)) {
            props.onClick(evt);
        }
    };

    let rootElement: JSX.Element;
    const extendedClassNames: string = classNames(className, CLASSNAME, { [`${CLASSNAME}--is-splitted`]: isSplitted });

    if (isSplitted) {
        rootElement = (
            <LxButtonGroup className={extendedClassNames}>
                <LxButton {...props} variant={Variants.button}>
                    {newChildren}
                </LxButton>

                <LxIconButton {...props} onClick={openDropdown}>
                    <LxIcon icon={mdiMenuDown} />
                </LxIconButton>
            </LxButtonGroup>
        );
    } else {
        rootElement = (
            <LxButton className={extendedClassNames} {...props} variant={Variants.button} onClick={openDropdown}>
                {newChildren}
                <LxIcon icon={mdiMenuDown} />
            </LxButton>
        );
    }

    return (
        <Fragment>
            {rootElement}

            {isDropdownOpened ? <LxDropdown>{dropdown}</LxDropdown> : null}
        </Fragment>
    );
};
LxDropdownButton.displayName = COMPONENT_NAME;

/////////////////////////////

export {
    CLASSNAME,
    DEFAULT_PROPS,
    Color,
    Colors,
    Emphasis,
    Emphasises,
    LxDropdownButton,
    LxDropdownButtonProps,
    Size,
    Sizes,
    Theme,
    Themes,
    Variant,
    Variants,
};
