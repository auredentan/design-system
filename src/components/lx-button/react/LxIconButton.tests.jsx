import React from 'react';
import { shallow } from 'enzyme';

import { LxIconButton } from './LxIconButton';
import { LxIcon } from '../../lx-icon/react/LxIcon';

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param  {Object} propOverrides An object that will extend the default properties.
 * @return {Object} An object with some shortcuts to elements or data required in tests.
 */
const setup = (propOverrides) => {
    const props = {
        ...propOverrides,
    };

    const mdiPlus = 'M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z';

    const wrapper = shallow(
        <LxIconButton {...props}>
            <LxIcon icon={mdiPlus} />
        </LxIconButton>,
    );

    return {
        props,
        wrapper,
    };
};

describe('<LxIconButton />', () => {
    // 1. Test render via snapshot (default state of component).
    it('should render correctly LxIconButton component', () => {
        const { wrapper } = setup();
        expect(wrapper).toMatchSnapshot();
    });

    // 2. Test defaultProps value and important props custom values.
    // N/A.

    // 3. Test events.
    // N/A.

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    // N/A.

    // 5. Test state.
    // N/A.
});
