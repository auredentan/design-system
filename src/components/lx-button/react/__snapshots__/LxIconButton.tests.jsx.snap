// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LxIconButton /> should render correctly LxIconButton component 1`] = `
<LxButton
  className="lx-button--shape-circled"
>
  <LxIcon
    className="lx-button__icon"
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".0"
  />
</LxButton>
`;
